qutip/solver/dysolve_propagator.py

```diff
@@ -0,0 +1,390 @@
+from qutip import Qobj
+import numpy as np
+import itertools
+from scipy.special import factorial
+
+__all__ = ['DysolvePropagator', 'dysolve_propagator']
+
+
+class DysolvePropagator:
+    """
+    A generator of propagator(s) using the Dysolve algorithm
+    (see https://arxiv.org/abs/2012.09282).
+
+    Parameters
+    ----------
+    max_order : int
+        The maximum order of approximation for the time evolution.
+        The bigger this variable is, the more terms are calculated,
+        so more accuracy but a longer time to compute the propagator(s).
+
+    H_0 : Qobj
+        The hamiltonian of the system.
+
+    X : Qobj
+        A cosine perturbation applied on the system.
+
+    omega : float
+        The frequency of the cosine perturbation.
+
+    a_tol : float, default: 1e-10
+        The absolute tolerance when it comes to say if values that
+        are computed are small enough to be considered as 0.
+
+    Notes
+    -----
+    The system's hamiltonian must be of the form
+    H = H_0 + cos(omega*t)X for Dysolve to work.
+
+    """
+
+    def __init__(
+        self,
+        max_order: int,
+        H_0: Qobj,
+        X: Qobj,
+        omega: float,
+        a_tol: float = 1e-10,
+    ):
+        self.max_order = max_order
+        self.H_0 = H_0
+        self.eigenenergies = H_0.eigenenergies()
+        self.X = X.transform(H_0)
+        self.omega = omega
+        self.t_i = None
+        self.t_f = None
+        self.dt = None
+        self.times = None
+        self.a_tol = a_tol
+        self._Sns = None
+        self.Us = None
+
+    def __call__(self, t_i: float, t_f: float, dt: float):
```

Comment:
This method should return the propagator at the given time, or between the given times:
```
U = DysolvePropagator(...)
psi(t=2) = U(2) @ psi(t=0)
psi(t=4) = U(4, 2) @ psi(t=2)
psi(t=0) = U(0, 4) @ psi(t=4)
U(t, t) == qeye(N)
```